src/engine/Server.cpp

```diff
@@ -665,20 +707,155 @@ Awaitable<void> Server::sendStreamableResponse(
+  }
+  AD_CONTRACT_CHECK(mediaType.has_value());
+  return mediaType.value();
+}
+
+// ____________________________________________________________________________
+Awaitable<void> Server::processQuery(
     const ad_utility::url_parser::ParamValueMap& params, const string& query,
     ad_utility::Timer& requestTimer,
     const ad_utility::httpUtils::HttpRequest auto& request, auto&& send,
     TimeLimit timeLimit) {
-  using namespace ad_utility::httpUtils;
+  MediaType mediaType = determineMediaType(params, request);
+  LOG(INFO) << "Requested media type of result is \""
+            << ad_utility::toString(mediaType) << "\"" << std::endl;
+
+  // TODO<c++23> use std::optional::transform
+  std::optional<uint64_t> maxSend = std::nullopt;
+  auto parameterValue =
+      ad_utility::url_parser::getParameterCheckAtMostOnce(params, "send");
+  if (parameterValue.has_value()) {
+    maxSend = std::stoul(parameterValue.value());
+  }
+  // Limit JSON requests by default
+  if (!maxSend.has_value() && (mediaType == MediaType::sparqlJson ||
+                               mediaType == MediaType::qleverJson)) {
+    maxSend = MAX_NOF_ROWS_IN_RESULT;
+  }
 
-  auto sendJson =
-      [&request, &send](
-          const json& jsonString,
-          http::status responseStatus) -> boost::asio::awaitable<void> {
-    auto response = createJsonResponse(jsonString, request, responseStatus);
-    co_return co_await send(std::move(response));
-  };
+  auto queryHub = queryHub_.lock();
+  AD_CORRECTNESS_CHECK(queryHub);
+  ad_utility::websocket::MessageSender messageSender{getQueryId(request, query),
+                                                     *queryHub};
+
+  auto [cancellationHandle, cancelTimeoutOnDestruction] =
+      setupCancellationHandle(messageSender.getQueryId(), timeLimit);
+
+  // Do the query planning. This creates a `QueryExecutionTree`, which will
+  // then be used to process the query.
+  //
+  // NOTE: This should come after determining the media type. Otherwise, it
+  // might happen that the query planner runs for a while (recall that it many
+  // do index scans) and then we get an error message afterward that a
+  // certain media type is not supported.
+  auto [pinSubtrees, pinResult] = determineResultPinning(params);
+  LOG(INFO) << "Processing the following SPARQL query:"
+            << (pinResult ? " [pin result]" : "")
+            << (pinSubtrees ? " [pin subresults]" : "") << "\n"
+            << query << std::endl;
+  QueryExecutionContext qec(index_, &cache_, allocator_,
+                            sortPerformanceEstimator_, std::ref(messageSender),
+                            pinSubtrees, pinResult);
+  auto plannedQuery = co_await setupPlannedQuery(
+      params, query, qec, cancellationHandle, timeLimit, requestTimer);
+  auto qet = plannedQuery.queryExecutionTree_;
+
+  if (plannedQuery.parsedQuery_.hasUpdateClause()) {
+    throw std::runtime_error("Expected Query but received Update.");
+  }
```

Comment:
I think this indicates a programming bug if true, so it can be AD_CORRECTNESS_CHECK etc. (which can also have a message).

Reply:
Currently this is not a programming bug but a standard violation. At the moment it is possible to pass in updates as requests that should contain queries and vice versa. When we split these up again in the code and also the grammar only then does this become a programming bug imo.